### ISSUE TITLE
Request#withPathInfo deletes PathInfoCaret

### ISSUE THREAD

**Issue by nicktelford** [http4s/http4s#2726]
In a `Request`, the `uri.path` for the request is split in to two parts, `scriptName` and `pathInfo` using the `Request.Keys.PathInfoCaret`. This is used for example by the `Router` middleware, which moves the caret such that the prefix is removed from `pathInfo`, but still accessible via `scriptName`.

When a user (e.g. another middleware) calls `withPathInfo`, my expectation is that it would only replace the `pathInfo` of the current request, and leave the `scriptName` untouched. However, this method calls `withUri`, which deletes the `PathInfoCaret`.

More formally, I would expect that:
```
val r = req.withPathInfo(x)
r.scriptName == req.scriptName
r.pathInfo = x
```

**Comment by rossabaker** [comment]
I share that expectation.  In fact, this bug is achingly familiar.  Let's fix it and put a better test on it.

**Comment by hamnis** [comment]
This was fixed by #2728
